apps/editing-toolkit/editing-toolkit-plugin/premium-content/subscription-service/class-token-subscription-service.php

```diff
@@ -178,12 +219,38 @@ protected function validate_subscriptions( $valid_plan_ids, $token_subscriptions
 	}
 
 	/**
-	 * @param  int    $site_id
-	 * @param  string $redirect_url
-	 * @return string
+	 * Get the URL of the JWT endpoint.
+	 *
+	 * @param  int    $site_id Site ID.
+	 * @param  string $redirect_url URL to redirect after checking the token validity.
+	 * @return string URL of the JWT endpoint.
 	 */
 	private function get_rest_api_token_url( $site_id, $redirect_url ) {
-		return sprintf( '%smemberships/jwt?site_id=%d&redirect_url=%s', self::REST_URL_ORIGIN, $site_id, urlencode( $redirect_url ) );
+		return sprintf( '%smemberships/jwt?site_id=%d&redirect_url=%s', self::REST_URL_ORIGIN, $site_id, rawurlencode( $redirect_url ) );
 	}
 
+	/**
+	 * Report the subscriptions as an ID => [ 'end_date' => ]. mapping
+	 *
+	 * @param array $subscriptions_from_bd List of subscriptions from BD.
+	 *
+	 * @return array<int, array>
+	 */
+	public static function abbreviate_subscriptions( $subscriptions_from_bd ) {
+		$subscriptions = array();
+		foreach ( $subscriptions_from_bd as $subscription ) {
+			// We are picking the expiry date that is the most in the future.
+			if (
+				'active' === $subscription['status'] && (
+					! isset( $subscriptions[ $subscription['product_id'] ] ) ||
+					empty( $subscription['end_date'] ) || // Special condition when subscription has no expiry date - we will default to a year from now for the purposes of the token.
```

Comment:
What's this special condition?

Reply:
There are plans that don't have an end time subscription (one-time payment plans), but `validate_subscriptions` need all subscriptions to have an end time set:

https://github.com/Automattic/wp-calypso/blob/51ceddf9982f5ff32de221b1f01107e56c152314/apps/editing-toolkit/editing-toolkit-plugin/premium-content/subscription-service/class-token-subscription-service.php#L212

So for these plans we set a fictional end date of a year for now so they are always considered valid and active by `validate_subscriptions`.